internal/provider/vaultradar/resource_radar_source_github_enterprise.go

```diff
@@ -0,0 +1,235 @@
+	GitHubOrganization types.String `tfsdk:"github_organization"`
+	Token              types.String `tfsdk:"token"`
+	ProjectID          types.String `tfsdk:"project_id"`
+}
+
+func (r *sourceGitHubEnterpriseResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	r.client = client
+}
+
+func (r *sourceGitHubEnterpriseResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	modifiers.ModifyPlanForDefaultProjectChange(ctx, r.client.Config.ProjectID, req.State, req.Config, req.Plan, resp)
+}
+
+func (r *sourceGitHubEnterpriseResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan gitHubEnterpriseSource
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectID := r.client.Config.ProjectID
+	if !plan.ProjectID.IsUnknown() {
+		projectID = plan.ProjectID.ValueString()
+	}
+
+	res, err := clients.OnboardRadarSource(ctx, r.client, projectID, radar_service.OnboardDataSourceBody{
+		Type:          "github_enterprise",
+		Name:          plan.GitHubOrganization.ValueString(),
+		ConnectionURL: plan.DomainName.ValueString(),
+		Token:         plan.Token.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating Radar source", err.Error())
+		return
+	}
+
+	plan.ID = types.StringValue(res.GetPayload().ID)
+	plan.ProjectID = types.StringValue(projectID)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+
+	tflog.Trace(ctx, "Created Radar resource")
+}
+
+func (r *sourceGitHubEnterpriseResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state gitHubEnterpriseSource
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectID := r.client.Config.ProjectID
+	if !state.ProjectID.IsUnknown() {
+		projectID = state.ProjectID.ValueString()
+	}
+
+	res, err := clients.GetRadarSource(ctx, r.client, projectID, state.ID.ValueString())
+	if err != nil {
+		if clients.IsResponseCodeNotFound(err) {
+			// Resource is no longer on the server.
+			tflog.Info(ctx, "Radar source not found, removing from state.")
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError("Unable to get Radar source", err.Error())
+		return
+	}
+
+	// Resource is marked as deleted on the server.
+	if res.GetPayload().Deleted {
+		tflog.Info(ctx, "Radar source marked for deletion, removing from state.")
+		resp.State.RemoveResource(ctx)
+		return
+	}
```

Comment:
I thought we want to remove the resource from state only if it is deleted. If we remove it while it is being deleted, could there be any side-effects? I am wondering about a case where the user runs `terraform destroy` and the deletion fails but we remove it from the state when they run `terraform plan` again

Reply:
Make sense, I sync with you later but I think the best thing for this case then is not to update state, but maybe just log a warning.